go/workflow-interceptor-updates.md

```diff
@@ -0,0 +1,417 @@
+
+```go
+InterceptWorkflow(ctx Context, next WorkflowInboundCallsInterceptor) WorkflowInboundCallsInterceptor
+```
+
+Why: Callers can use the context to get workflow info in addition to other things (e.g. loggers, metricsm, etc). Also
+using the context as the first parameter is good Go practice and allows for more contextual information to be provided
+in the future.
+
+### WorkflowXCallsInterceptor
+
+Proposal: Change the name to `WorkflowXInterceptor`. Why: `Calls` just adds noise (more accepted in Java, not Go).
+
+### WorkflowInboundCallsInterceptor.ExecuteWorkflow
+
+Current:
+
+```go
+ExecuteWorkflow(ctx Context, workflowType string, args ...interface{}) []interface{}
+```
+
+Proposal:
+
+```go
+type WorkflowInput struct {
+  Args []interface{}
+}
+
+ExecuteWorkflow(ctx Context, input *WorkflowInput) (interface{}, error)
+```
+
+Why:
+
+* Workflow type can be derived from the context _or_ the info first given to `InterceptWorkflow`
+* Workflows can only return a value + error or just an error and this makes it more explicit what the error is
+
+### WorkflowInboundCallsInterceptor.ProcessSignal
+
+Current:
+
+```go
+ProcessSignal(ctx Context, signalName string, arg interface{}) error
+```
+
+Proposal:
+
+```go
+type SignalInput struct {
+  SignalName string
+  Arg interface{}
+}
+
+HandleSignal(ctx Context, input *SignalInput) error
+```
+
+Also, this will be called to handle the signal with the last invocation in the chain passing the signal to the workflow.
+
+Why:
+
+* `HandleSignal` matches Java nomenclature and new behavior better
+* Currently the sending of the signal to the handler is done after the interceptor not allowing the interceptor to
+  actually intercept and alter/prevent the send. We will change to have the last interceptor do the actual send.
+
+### WorfkflowInboundCallsInterceptor.HandleQuery
+
+Current:
+
+```go
+HandleQuery(ctx Context, queryType string, args *commonpb.Payloads,
+  handler func(*commonpb.Payloads) (*commonpb.Payloads, error)) (*commonpb.Payloads, error)
+```
+
+Proposal:
+
+```go
+type QueryInput struct {
+  QueryType string
+  Args []interface{}
+}
+
+HandleQuery(ctx Context, input *QueryInput) (interface{}, error)
+```
+
+Why:
+
+* All other calls convert the payload outside of the interceptor call, this can too
+* No need to pass a handler, that is just part of the last item in the interceptor chain (i.e. when calling next)
+
+### WorkflowOutboundCallsInterceptor
+
+We can leave everything the way it is because, while it doesn't make the interceptor parameters future-proof with struct
+wrappers, it mimics the actual calls that aren't going to have backwards-compatible alterations.
+
+### WorkflowOutboundCallsInterceptor.GetWorkflowInfo
+
+Current:
+
+```go
+GetWorkflowInfo(ctx Context) *WorkflowInfo
+```
+
+Proposal:
+
+```go
+GetInfo(ctx Context) *WorkflowInfo
+```
+
+Why: All other calls match `workflow` package, why not this?
+
+### WorkflowOutboundCallsInterceptor.NewContinueAsNewError
```

Comment:
Per discussion, removing this since it is not the best way to capture continue as new